### ISSUE TITLE
Generating strings with Instaparse

### ISSUE THREAD

**Issue by zmaril** [Engelberg/instaparse#82]
I'm interested in generating strings with instaparse. By this, I mean I'd like to give instaparse a data structure to fill in and a grammar and have it return to me strings that could parse to something that would parse to the given data structure. We use instaparse a good deal in [ECHELON](https://github.com/sunlightlabs/echelon) and I'd like to use `test.check` to generate data for testing. I'm still learning how the QuickCheck style libraries work, but I think if we could generate strings via instaparse we could use `test.check` without too much trouble. I'm happy to develop this myself (and will continue to explore further), but I wanted to check in first to see if it is possible (or has already been quietly done). 

For clarification, I'd like a function that does something akin to the following: 

``` clj
(def grammar
  (insta/parser
   "S = name ' ' business
    name = 'Steve' | 'Mary' | 'Bob'
    business = 'Bakery' | 'Shop'"))

(grammar "Steve Shop")
(def grammar
  (insta/parser
   "S = name ' ' business
    name = 'Steve' | 'Mary' | 'Bob'
    business = 'Bakery' | 'Shop'"))

(grammar "Steve Shop")

(insta/generate [:S [:name] [:business]])
;;;[:S [:name "Steve"] " " [:business "Shop"]]

(insta/generate [:S [:name] [:business]])
;;["Steve Bakery" "Steve Shop" "Mary Bakery" "Mary Shop" "Bob Bakery" "Bob Shop"]
```

I suppose I'd like to give instaparse a partial data structure corresponding to a "skeleton" of a parse tree and have it generate data to fill in the rest of the tree. I haven't thought through how this would play with transformations and the hiding of tags in instaparse yet. Does this seem possible? Is it a dumb thing to ask for?

**Comment by kanaka** [comment]
As part of research I'm working on for my PhD (browser render testing), I recently created [instacheck](https://github.com/kanaka/instacheck) which is basically instaparse meets test.check. It allows you to define EBNF grammars (instaparse compatible) which are translated into test.check generators. It has some command line modes for generating samples, running test cases/inputs against a test program (with shrinking), and also generating the translated clj generator code for a grammar. It even supports regex in the grammar thanks to the awesome test.chuck from @gfredericks. I just got back from presenting the project at Strange Loop. I think videos should post within the next week.

One critical feature that instacheck supports is the ability to adjust the weights that are used for alternation/choice points in the grammar. This is super useful and probably critical for a lot of applications (although this wasn't obvious until I started using it in anger).

Note that for the moment, instacheck requires patched version of test.check and instaparse. The test.check change to to get the frequency generator to shrink towards the more common weights rather than shrinking to whatever happens to be defined first. I should be able to get rid of this requirement by defining my own frequency generator that basically does weight sorting prior to calling the underlying test.check frequency generator. The changes I have to instaparse are to support the ability to tune the grammar weights: https://github.com/Engelberg/instaparse/pull/183, https://github.com/Engelberg/instaparse/pull/184. The first allows weights to be configured via comments embedded directly in the grammar. The second patch enables logging of the paths that are taken through a grammar. This allows you to tune weights according to existing test cases by parsing them with the grammar.

**Comment by zmaril** [comment]
Wow! That's awesome to see. Way to go, that's seriously cool, thank you for posting about it on this thread.

**Comment by Engelberg** [comment]
Great work! Now that I understand the purpose of the pull requests, it will be easier to evaluate those. Thanks for this.

**Comment by arrdem** [comment]
!voteclose ;)

**Comment by lvh** [comment]
I would like to generate a _specific_ string from a _speciifc_ parse tree (given a grammar, of course). Is this the right ticket for that?

**Comment by gfredericks** [comment]
@lvh I'm not sure what you mean by that; you already have the string? and the parse tree? what "generation" is there left to do? maybe an example would help

**Comment by lvh** [comment]
@gfredericks Right, the idea is you'd either generate a parse tree from scratch, or more likely modify a parse tree and then reserialize. So you can produce a semantic patch () but render it as a diff, for example.

**Comment by gfredericks** [comment]
are you just saying you want to unparse a tree? for some grammars there would be more than a single valid output

**Comment by lvh** [comment]
Sorry, lost track of this. Yes, that's what I'm saying. I'd be happy with any output; also happy to make the grammar capture more details of the input file to make the output as close as possible.
